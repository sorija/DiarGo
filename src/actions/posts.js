import database from '../firebase/firebase';
import moment from 'moment';

export const addPost = (post) => ({
  type: 'ADD_POST',
  post
});

export const startAddPost = (postData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      content = '',
      createdAt = 0
    } = postData;
    // save data
    const post = { title, content, createdAt };
    return database.ref(`users/${uid}/posts`).push(post).then((ref) => {
      dispatch(addPost({
        id: ref.key,
        ...post
      }));
    });
  };
};

export const setPosts = (posts) => ({
  type: 'SET_POSTS',
  posts
});

export const startSetPosts = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts`).once('value').then((snapshot) => {
      const posts = [];
      snapshot.forEach((childSnapshot) => {
        posts.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setPosts(posts));
    });
  };
};

export const removePost = ({ id } = {}) => ({
  type: 'REMOVE_POST',
  id
});

export const startRemovePost = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts/${id}`).remove().then(() => {
      dispatch(removePost({ id }));
    });
  };
};

export const editPost = (id, updates) => ({
  type: 'EDIT_POST',
  id,
  updates
});

export const startEditPost = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts/${id}`).update(updates).then(() => {
      dispatch(editPost(id, updates));
    });
  };
};
