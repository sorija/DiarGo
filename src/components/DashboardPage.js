import React from 'react';
import { Link } from 'react-router-dom';
import PostList from './PostList';
import PostListFilters from './PostListFilters';

const DashboardPage = () => (
  <div>
    <Link className="button" to="/create">New post</Link>
    <PostListFilters />
    <PostList />
  </div>
);

export default DashboardPage;
