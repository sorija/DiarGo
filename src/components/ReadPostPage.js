import React from 'react';
import { connect } from 'react-redux';

export const ReadPostPage = (props) => (
  <div>
    <h1>Post Page</h1>
    <h3>{props.post.title}</h3>
    <p>{props.post.content}</p>
  </div>
);

const mapStateToProps = (state, props) => ({
  post: state.posts.find((post) => post.id === props.match.params.id)
});

export default connect(mapStateToProps)(ReadPostPage);
