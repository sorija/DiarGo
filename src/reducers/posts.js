const postsReducer = (state = [], action) => {
  switch (action.type) {
    case 'ADD_POST':
      return [...state, action.post];
    case 'SET_POSTS':
      return action.posts;
    case 'EDIT_POST':
    return state.map((post) => {
      if (post.id === action.id) {
        return {
          ...post,
          ...action.updates
        };
      } else {
        return post;
      };
    });
    case 'REMOVE_POST':
      return state.filter(({ id }) => id !== action.id);
    default:
      return state;
  }
};

export default postsReducer;
