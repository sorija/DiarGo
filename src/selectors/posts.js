import moment from 'moment';

const selectPosts = (posts, { text, date }) => {
  return posts.filter((post) => {
    const createdAtMoment = moment(post.createdAt);
    const textMatch = post.title.toLowerCase().includes(text.toLowerCase());
    const dateMatch = date ? date.isSame(createdAtMoment, 'day') : true;
    
    return textMatch && dateMatch;
  });
};

export default selectPosts;
