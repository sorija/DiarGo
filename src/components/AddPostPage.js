import React from 'react';
import { connect } from 'react-redux';
import PostForm from './PostForm';
import { startAddPost } from '../actions/posts';

export class AddPostPage extends React.Component {
  onPostSubmit = (post) => {
    this.props.startAddPost(post);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add New Post</h1>
          </div>
        </div>
        <div className="content-container">
          <PostForm 
            onPostSubmit={this.onPostSubmit}
          />
        </div>
      </div>
    );
  };
};

const mapDispatchtoProps = (dispatch) => ({
  startAddPost: (post) => dispatch(startAddPost(post))
});

export default connect(undefined, mapDispatchtoProps)(AddPostPage);
