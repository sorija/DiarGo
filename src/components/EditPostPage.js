import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PostForm from './PostForm';
import { startEditPost, startRemovePost } from '../actions/posts';

export class EditPostPage extends React.Component {
  onSubmit = (post) => {
    this.props.startEditPost(this.props.post.id, post)
    this.props.history.push('/');
  };
  onRemove = () => {
    this.props.startRemovePost({ id: this.props.post.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <Link to={`/read/${this.props.post.id}`}>Readable version</Link>
        <PostForm
          post={this.props.post}
          onPostSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove Post</button>
      </div>
    );
  };
};

const mapStateToProps = (state, props) => ({
  // search state for matching post
  post: state.posts.find((post) => post.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditPost: (id, post) => dispatch(startEditPost(id, post)),
  startRemovePost: (data) => dispatch(startRemovePost(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPostPage);
