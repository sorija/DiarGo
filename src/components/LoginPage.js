import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">DiarGo</h1>
      <p>Your personal web diary</p>
      <button className="button button--log" onClick={startLogin}>Login with Google</button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
