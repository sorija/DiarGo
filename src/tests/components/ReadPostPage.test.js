import React from 'react';
import { shallow } from 'enzyme';
import { ReadPostPage } from '../../components/ReadPostPage';
import posts from '../fixtures/posts';

test('should render ReadPostPage with a post', () => {
  const wrapper = shallow(<ReadPostPage post={posts[0]}/>);
  expect(wrapper).toMatchSnapshot();
});
