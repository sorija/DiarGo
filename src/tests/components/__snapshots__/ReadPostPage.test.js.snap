// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ReadPostPage with a post 1`] = `
<div>
  <h1>
    Post Page
  </h1>
  <h3>
    Test Post Title
  </h3>
  <p>
    Test Post Content
  </p>
</div>
`;
